contracts/router/tree/TokenTree.sol

```diff
@@ -0,0 +1,578 @@
+        }
+        // First, we traverse up the tree from "tokenFrom" to one level deeper the lowest common ancestor.
+        route = _multiSwapToRoot(0, rootPathFrom, req.from.depth, depthDiff, amountIn);
+        // Check if we need to do a sibling swap. When the two nodes are connected to the same parent via the same pool,
+        // we do a direct swap between the two nodes, instead of doing two swaps through the parent using the same pool.
+        uint256 lastNodeIndex = _extractNodeIndex(rootPathFrom, depthDiff);
+        uint256 siblingIndex = _extractNodeIndex(rootPathTo, depthDiff);
+        uint256 firstPoolIndex = _nodes[lastNodeIndex].poolIndex;
+        uint256 secondPoolIndex = _nodes[siblingIndex].poolIndex;
+        if (firstPoolIndex == secondPoolIndex) {
+            // Swap lastNode -> sibling
+            (route.visitedPoolsMask, route.amountOut) = _singleSwap(
+                route.visitedPoolsMask,
+                firstPoolIndex,
+                lastNodeIndex,
+                siblingIndex,
+                route.amountOut
+            );
+        } else {
+            // Swap lastNode -> parent
+            uint256 parentIndex = _extractNodeIndex(rootPathFrom, depthDiff - 1);
+            (route.visitedPoolsMask, route.amountOut) = _singleSwap(
+                route.visitedPoolsMask,
+                firstPoolIndex,
+                lastNodeIndex,
+                parentIndex,
+                route.amountOut
+            );
+            // Swap parent -> sibling
+            (route.visitedPoolsMask, route.amountOut) = _singleSwap(
+                route.visitedPoolsMask,
+                secondPoolIndex,
+                parentIndex,
+                siblingIndex,
+                route.amountOut
+            );
+        }
+        // Finally, we traverse down the tree from the lowest common ancestor to "tokenTo".
+        return _multiSwapFromRoot(route.visitedPoolsMask, rootPathTo, depthDiff, req.to.depth, route.amountOut);
+    }
+
+    /// @dev Performs a multi-hop swap, going in "from root direction" (where depth increases)
+    /// via the given `rootPath` from `depthFrom` to `depthTo`.
+    /// Assumes that the initial token is already in this contract.
+    function _multiSwapFromRoot(
+        uint256 visitedPoolsMask,
+        uint256 rootPath,
+        uint256 depthFrom,
+        uint256 depthTo,
+        uint256 amountIn
+    ) internal returns (Route memory route) {
+        uint256 nodeIndex = _extractNodeIndex(rootPath, depthFrom);
+        // Traverse down the tree following `rootPath` from `depthFrom` to `depthTo`.
+        for (uint256 depth = depthFrom; depth < depthTo; ) {
+            // Get the child node
+            unchecked {
+                ++depth;
+            }
+            uint256 childIndex = _extractNodeIndex(rootPath, depth);
+            // Swap node -> child
+            (visitedPoolsMask, amountIn) = _singleSwap(
```

Comment:
Using `amountOut = amountIn` from e.g.

```
IDefaultPool(pool).swap({
                tokenIndexFrom: tokenIndexes[pool][tokenFrom],
                tokenIndexTo: tokenIndexes[pool][tokenTo],
                dx: amountIn,
                minDy: 0,
                deadline: type(uint256).max
            });
```

might not be the *actual* balance received from the pool on each swap due to non-standard transfer fees (e.g. USDT).

Consider doing a balance update on `amountOut`/`dy` either here or probably easier/more general in L133 of `_singleSwap`

Reply:
Makes sense and understand the point. I think probably better to be safe here though with the balance checks, as it's possible someone forgets down the line and deploys with a token that has transfer fees, even if unlikely.

The alternative, to be explicit, would be to do the change at `LinkedPool` level in the individual [`_poolSwap`](https://github.com/synapsecns/synapse-contracts/blob/feat/universal-swap-wrapper/contracts/router/LinkedPool.sol#L253) implementation, then have in this same repo a `LinkedPoolWithTransferFees` that does the balance checks (although would vouch for balance checks in one contract to keep it simple)

Reply:
Another example where this *may* come up and be important would be rebase tokens, like stETH (if looking to support this down the line). I need to look back and check how the rebase is done exactly, but could be an issue if using in the stable swap pool (like Curve's stETH pool for bridging)

Reply:
Ended up ignoring potential transfer fees here: [8fc94a8](https://github.com/synapsecns/synapse-contracts/pull/229/commits/8fc94a81bbd4550e05633c75a1d706c14f065f8e).

As for the rebasing tokens, if we ever want to add the support for them, I feel this could be done via a dedicated module/adapter depending on type of integration.